File: public/javascripts/ec/game/game_commands/move_unit.js
var MoveUnit = new JS.Class(GameCommand, {
  initialize: function(game, user, params) {
    this.callSuper();

    this.unit_location = [this.params.unit_x, this.params.unit_y, this.params.unit_slot];
    this.dest_x = this.params.dest_x;
    this.dest_y = this.params.dest_y;
  },

  class_name: function() {
    return 'MoveUnit';
  },

  execute: function() {
    var unit = this.game.unit_at(this.unit_location[0], this.unit_location[1], this.unit_location[2]);

    var cost_map = new JS.Hash(null);

    for (var x = 0; x < this.game.map.tiles_width; x++) {
      for (var y = 0; y < this.game.map.tiles_height; y++) {
        var coords = new HCoordinate(x, y);
        var cost = unit.terrain_cost(this.game.terrain_at(x, y), this.game.unmodified_terrain_at(x, y));
        cost_map.put(coords, cost);
      }
    }

    var pf = new PathFinder(cost_map, unit, this.game.enemy_units(this.user));
    var remaining_mv = pf.possible_destination_tiles().get(new HCoordinate(this.dest_x, this.dest_y));

    this.move_unit(this.unit_location[0], this.unit_location[1], this.unit_location[2], this.dest_x, this.dest_y, null);
    this.modify(unit, 'movement_points', remaining_mv);
    this.modify(unit, 'moved', true);

    if (unit.def.attack_type == 'exclusive') {
      this.modify(unit, 'attacks', unit.def.attack_phases);
    }
  },

  can_undo: function() {
    return true;
  }
});

File: app/game_commands/move_unit.rb
class MoveUnit < GameCommand
  def initialize(game, user, params)
    super(game, user, params)

    @unit_location = @params.values_at(:unit_x, :unit_y, :unit_slot)
    @dest_x, @dest_y = *@params.values_at(:dest_x, :dest_y)
  end

  def execute
    require_current_player
    require_running_game
    unit = require_current_player_unit(*@unit_location)

    raise CommandError.new("Unit is capturing a base.") if @game.capturing_at?(*@unit_location[0..1])
    raise CommandError.new("No movement points.") if unit.movement_points == 0

    remaining_mv = require_destination_reachable(unit, @dest_x, @dest_y)

    move_unit(*@unit_location, @dest_x, @dest_y, nil)
    modify(unit, :movement_points, remaining_mv)
    modify(unit, :moved, true)

    if unit.attack_type == :exclusive
      modify(unit, :attacks, unit.attack_phases)
    end

    { :remaining_movement_points => remaining_mv }
  end

  protected

  def require_destination_reachable(unit, x, y)
    zoc_tiles = @game.rival_units.select do |u|
      u.can_zoc_unit_type?(unit.unit_type)
    end.map do |u|
      [u.x, u.y]
    end

    cost_map = {}

    @game.map.tiles_hash.each do |coords, tile_index|
      cost_map[coords] = unit.terrain_cost(@game.terrain_at(*coords), @game.unmodified_terrain_at(*coords))
    end

    pf = PathFinder.new(
      cost_map, unit, @game.rival_units
    )
    occupied_tiles = @game.units.map { |u| [u.x, u.y] } - [x, y]

    if unit.armor_type == :naval
      @game.terrain_modifiers.each do |tm|
        if tm.terrain_name == 'bridge'
          occupied_tiles += [tm.x, tm.y]
        end
      end
    end

    pf.possible_destination_tiles.except(*occupied_tiles)[[x, y]].tap do |remaining_mv|
      raise CommandError.new("Invalid destination.") unless remaining_mv
    end
  end
end
